bindings/rust/wasmedge-rs/src/vm.rs

```diff
@@ -1,98 +1,79 @@
-        };
-        if ctx.is_null() {
-            return Err(Error::Create);
-        }
-
-        unsafe {
-            ffi::decode_result(ffi::WasmEdge_VMLoadWasmFromASTModule(ctx, module.ctx))
-                .map_err(Error::ModuleLoad)?;
-            // The following two calls could be made lazily if they're expensive.
-            ffi::decode_result(ffi::WasmEdge_VMValidate(ctx)).map_err(Error::Validate)?;
-            ffi::decode_result(ffi::WasmEdge_VMInstantiate(ctx)).map_err(Error::Instantiate)?;
-        }
-
-        Ok(Self { ctx })
+    pub fn builder(module_path: &str) -> VmBuilder {
+        VmBuilder::new(module_path)
     }
 
-    pub fn run(
-        &mut self,
-        func_name: impl AsRef<str>,
-        params: &[Value],
-    ) -> Result<Vec<Value>, Error> {
-        let raw_func_name: ffi::WasmEdge_String =
-            crate::string::StringRef::from(func_name.as_ref()).into();
-
-        let raw_params: Vec<_> = params
-            .as_ref()
-            .iter()
-            .copied()
-            .map(ffi::WasmEdge_Value::from)
-            .collect();
+    pub fn default(module_path: &str) -> Result<Vm, Error> {
+        let config = wasmedge::Config::default();
+        let module = Module::new(&config, module_path)?;
+        let vm_instance = wasmedge::Vm::create(&module.inner, &config)?;
+        Ok(Vm { config, module, inner: Some(vm_instance), _private: ()})
+    }
 
-        let func_type = unsafe { ffi::WasmEdge_VMGetFunctionType(self.ctx, raw_func_name) };
-        if func_type.is_null() {
-            return Err(Error::MissingFunction(func_name.as_ref().to_string()));
-        }
-        let num_returns = unsafe { ffi::WasmEdge_FunctionTypeGetReturnsLength(func_type) };
-        let mut returns = Vec::with_capacity(num_returns as usize);
+    fn create(
+        self, 
+        module: &wasmedge::Module,
+        config: &wasmedge::Config,
+    ) -> Result<Self, Error> {
+        let vm = wasmedge::Vm::create(module, config).map_err(VmError::Create)?;
+        let vm = vm.load_wasm_from_ast_module(module).map_err(VmError::ModuleLoad)?; 
+        let vm = vm.validate().map_err(VmError::Validate)?;
+        let vm = vm.instantiate().map_err(VmError::Instantiate)?;
+        Ok(vm)
+    }
 
-        unsafe {
-            ffi::decode_result(ffi::WasmEdge_VMExecute(
-                self.ctx,
-                raw_func_name,
-                raw_params.as_ptr(),
-                raw_params.len() as u32,
-                returns.as_mut_ptr(),
-                num_returns,
-            ))
-            .map_err(Error::Execute)?;
-            returns.set_len(num_returns as usize);
+    pub fn run(self, func_name: &str, params: &[wasmedge::Value]) -> Result<Vec<Value>, Error> {
+        match self.inner {
+            Some(vm) => {
+                vm.run(func_name, params).map_err(Error::Execute)?;
+            }
+            None => panic!("WasmEdge Vm can't run!")
```

Comment:
You are right to be concerned.
The idea here is this. 
I want to construct a chain call so that it's easy to insert `storage`-like information in subsequent modifications. Something like this.
``rust
let config = wedge::Config::default();
let storage = wedge::Storage::default();
let vm = wedge::Vm::new(module_path)?
                .with_config(config)?
                .with_storage(storage)?
                .apply(WASI)?  
                .build()? ;

```

Vm here is just a wrapper for the upper layer:

```rust

pub struct Vm{
    config: Option<Arc<wasmedge::Config>>,
    module: Option<Arc<Module>>,
    inner: Option<wasmedge::Vm>,

}
```

They are `None` when `Vm::new` is called, and they are Some after the C-API is actually called later and the vm instance is actually created. vm doesn't go run if it is just None.